### DISCUSSION TITLE
Script doesn't use GPU

### DISCUSSION THREAD

**Discussion by bjoernapfel** [genn-team/genn#586] [GeNN questions]
I run my pyGeNN script on ubuntu, despite following the steps V and VI in the GeNN installation process (and ensuring that my GPU, CUDA and everything is compatible and set up), my script somehow only seems to run over my CPU and doesn´t connect to the GPU - it's very slow and when checking with the top program, 100% of my CPU is being used for the process which shouldn´t be the case when actually using the GPU. Nothing changes upon un- and reinstalling. Is this maybe some sort of known error you have a simple solution for?
Maybe it helps clarifying my problem, when running the script I also get multiple messages like 
![image](https://github.com/genn-team/genn/assets/116563213/b31b0dea-4636-44ed-bf71-cbe74418a0a2)
![image](https://github.com/genn-team/genn/assets/116563213/2c091a0c-8f8e-4d2d-a3aa-e5c2e05860d2)

**Comment by neworderofjamie** [comment]
Did you set the ``CUDA_PATH`` environment variable before building PyGeNN?

**Comment by neworderofjamie** [comment]
You can also verify whether CUDA was correctly detected when building PyGeNN by running the following:
```python
import pygenn
print(pygenn.genn_model.backend_modules)
```
If the first entry in the dictionary doesn't have a key of "CUDA" then the CUDA backend was not build correctly

**Reply by bjoernapfel** [reply]
my output when running these lines:
![image](https://github.com/genn-team/genn/assets/116563213/878778c9-e790-450e-93d7-2d933d37db77)

**Reply by bjoernapfel** [reply]
so I don´t think it is connected to to CUDA not being build correctly

**Comment by bjoernapfel** [comment]
do you maybe have any other suggestions?

**Comment by neworderofjamie** [comment]
Apologies, I've been out of the office for a few days. Your output suggests that the CUDA backend was built and loaded correctly so it **should** be being used. Could you post the output from the ``nvidia-smi``  command line tool?

**Reply by bjoernapfel** [reply]
Sure, here it is
![image](https://github.com/genn-team/genn/assets/116563213/a2e374a1-663f-46c9-9c21-391db737a0dd)

**Reply by neworderofjamie** [reply]
OK, that looks perfectly healthy. Can you try running https://github.com/neworderofjamie/genn_examples/blob/master/pynn/potjans_microcircuit_pygenn.py and tell me the performance numbers you get?

**Reply by bjoernapfel** [reply]
![image](https://github.com/genn-team/genn/assets/116563213/566de929-84d7-4d90-9f5a-1a03e8651889)

**Reply by neworderofjamie** [reply]
Ok, so GeNN is working fine with CUDA on your GPU and the performance is about as I would expect for an A4000. Therefore the issue is probably with your model - could you share it with me? My email's J.C.Knight at sussex.ac.uk

**Reply by bjoernapfel** [reply]
Done!:)

**Reply by neworderofjamie** [reply]
Just had a quick try with your model and basically what you were seeing was the code generator generating vast amounts of code and the CPU stuck compiling it rather than the model running :smile: Sadly, mostly due to attempts to maintain backward compatibility, it's quite easy to get into this situation with GeNN. If you add the following lines somewhere before you build your model:

```python
model._model.set_fuse_postsynaptic_models(True)
model._model.set_fuse_pre_post_weight_update_models(True)
```

The model builds and simulates - look forward to seeing the results!

**Reply by bjoernapfel** [reply]
I now added the lines immediately after creating the model. For me, however, it still doesn´t build and simulate - instead I get the new error 
![image](https://github.com/genn-team/genn/assets/116563213/995323fa-491d-44ba-a6df-063bcf262072)
can you think of further genn code addressing this issue?

**Reply by neworderofjamie** [reply]
Can you make sure you are using the latest version of GeNN? This bug should have been fixed in the 4.8.1 release.

**Reply by bjoernapfel** [reply]
![image](https://github.com/genn-team/genn/assets/116563213/85f50e6e-d4ff-49fb-aa64-9854810757e6)
To my understanding, 4.8.1 is the latest version?

**Reply by neworderofjamie** [reply]
could you email me the zipped up .cc files from the ``Hypercolum_CODE`` directory so I can try and debug? Once you've done that, you can maybe workaround this by passing ``constantCacheOverhead=500`` to the ``GeNNModel`` constructor - this could maybe be a CUDA 12 thing.

**Reply by bjoernapfel** [reply]
Sorry the customUpdate.cc file got messed up, I got confused when shipping the files between the ubuntu machine and my local laptop. But either way, after adding the constantCacheOverhead=500 the model did actually build! I got performance times likes these, were they close to when you ran the model?
![image](https://github.com/genn-team/genn/assets/116563213/229f1027-d59b-424c-b6af-139994c413fe)

**Reply by neworderofjamie** [reply]
Hah I was wondering why customUpdate.cc had Python in :laughing: Glad my workaround succeeded and your code is building though! I suspect this is a CUDA 12 thing but, to save me having to upgrade my machine, could you remove the workaround and, after you get the constant cache error, zip up all the .o files in the ``Hypercolum_CODE`` directory and send them over?

**Reply by neworderofjamie** [reply]
and yes, on my A5000 I see:
```
Timing:
	Simulation:0.000429
	Init:0.839680
	Sparse init:0.034656
	Neuron simulation:714.322860
	Synapse simulation:608.035744
```
which is approximately the same (``sim_start_time`` isn't  getting measured before the simulation loop hence the total simulation time is very small)

**Reply by bjoernapfel** [reply]
done, thanks a lot!

**Reply by bjoernapfel** [reply]
Ok it actually seemed like your workaround only did it for these specific configurations, I upscaled the model just a bit from 4MC and 7HC to 4MC and 9HC (I’d like to eventually run simulations with up to 9MC and 9-12HC) and then I already got the almost same mistake again.
![image](https://github.com/genn-team/genn/assets/116563213/bb102d01-05bc-4dd6-a3f0-142e24e8237b)
Is it possible to update your workaround? I otherwise also sent you the .cc files of the recently failed simulation.

**Reply by neworderofjamie** [reply]
Interesting! I will investigate this further and fix ASAP. In the mean time, I imagine if you increase the ``constantCacheOverhead`` somewhat you'll find a point where it works

**Reply by neworderofjamie** [reply]
Also, if it's possible, you could try downgrading to CUDA 11.X - knowing whether this is a CUDA 12 issue or not would be very helpful

**Reply by bjoernapfel** [reply]
interestingly, after downgrading to CUDA 11.8, none of the above problems showed up and the script works just fine, thanks! so it actually appears to have been a CUDA 12 issue

**Reply by neworderofjamie** [reply]
For now downgrading to CUDA 11.X is probably the best workaround but clearly something is going awry with CUDA 12 which needs investigating - will raise a bug

**Reply by neworderofjamie** [reply]
Turns out this had nothing to do with CUDA versions, whether it occurred or not was dependent on the non-deterministic construction of this model. Nonetheless, it is fixed in https://github.com/genn-team/genn/pull/589
